File: bnovate/bnovate/report/material_request_report/material_request_report.js
// Copyright (c) 2016, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Material Request Report"] = {
    filters: [
        {
            fieldname: "request_type",
            label: __("Request Type"),
            fieldtype: "Select",
            options: ["", "Material Transfer", "Purchase", "Manufacture"],
        },
        {
            fieldname: "item_group",
            label: __("Item Group"),
            fieldtype: "Link",
            options: "Item Group",
        },
        {
            fieldname: "assigned_to",
            label: __("Assigned To"),
            fieldtype: "Link",
            options: "User",
        },
        {
            "fieldname": "only_unassigned",
            "label": __("Only Unassigned"),
            "fieldtype": "Check",
            "default": 0
        }
    ],
    formatter(value, row, col, data, default_formatter) {
        if (col.fieldname === "assigned_to") {
            // return `<a >${default_formatter(date, row, col, data)}</a>`;
            return `<a onclick="assign('${data.mr_item}', '${data.assigned_to}')">${value || '<i>Unassigned</i>'}</a>`;
        } else if (col.fieldname === 'purchase_order') {
            if (value) {

                let [legend, colour] = purchase_order_indicator(data);
                return `<span class="indicator ${colour}">${frappe.utils.get_form_link("Purchase Order", value, true, value)} [${legend}]</span>`;
            } else {
                return `<button class="btn btn-xs btn-primary create-po" onclick="create_po('${data.supplier}')">Create PO</button>`;
            }
        }
        return default_formatter(value, row, col, data);
    }
};

// colours borrowed from sales_order_list.js on ERPNext app
function purchase_order_indicator(doc) {
    colormap = {
        "Draft": "red",
        "Closed": "green",
        "On Hold": "orange",
        "Delivered": "green",
        "To Receive and Bill": "orange",
        "To Receive": "orange",
        "To Bill": "orange",
        "Completed": "green"
    }
    return [__(doc.po_status), colormap[doc.po_status]]
}


async function assign(mr_item, assigned_to) {

    const fields = [{
        fieldname: "assigned_to",
        fieldtype: "Link",
        options: "User",
        label: __("Assign To"),
        default: assigned_to == 'null' ? "" : assigned_to,

    }];

    let values = await bnovate.utils.prompt("Edit dates", fields, "Confirm", "Cancel")

    if (!values) {
        return;
    }

    await frappe.db.set_value("Material Request Item", mr_item, "assigned_to", values.assigned_to);

    frappe.query_report.refresh();
}

async function create_po(supplier) {

    // Create PO for all open MRs for the same supplier
    const resp = await frappe.call({
        method: "erpnext.stock.doctype.material_request.material_request.make_purchase_order_based_on_supplier",
        args: { source_name: supplier }
    });

    // const doc = frappe.get_doc(resp.message);
    const doc = resp.message;

    // Setting price list rate from BKO works, but amount is overriden by a link trigger - just let it be calculated on save.
    doc.items.forEach(i => {
        if (i.blanket_order) {
            i.price_list_rate = i.blanket_order_rate;
        }
    });
    frappe.model.sync(doc);
    frappe.set_route("Form", doc.doctype, doc.name);
}

File: bnovate/bnovate/report/material_request_report/material_request_report.py
# Copyright (c) 2013, libracore and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe

def execute(filters=None):
    return get_columns(), get_data(filters)

def get_columns():
    """
    Column definitions extracted from the SQL SELECT aliases.
    """
    return [
        {"fieldname": "material_request", "label": "Material Request", "fieldtype": "Link", "options": "Material Request", "width": 120},
        {"fieldname": "request_type", "label": "Request Type", "fieldtype": "Data", "width": 120},
        {"fieldname": "transaction_date", "label": "Date", "fieldtype": "Date", "width": 100},
        {"fieldname": "schedule_date", "label": "Required Date", "fieldtype": "Date", "width": 100},
        {"fieldname": "item_code", "label": "Item Code", "fieldtype": "Link", "options": "Item", "width": 220, "align": "left"},
        # {"fieldname": "description", "label": "Description", "fieldtype": "Data", "width": 200},
        {"fieldname": "supplier", "label": "Default Supplier", "fieldtype": "Link", "options": "Supplier", "width": 140},
        {"fieldname": "supplier_name", "label": "Supplier Name", "fieldtype": "Data", "width": 180},
        {"fieldname": "item_group", "label": "Item Group", "fieldtype": "Link", "options": "Item Group", "width": 140}, 
        {"fieldname": "assigned_to", "label": "Assigned To", "fieldtype": "Link", "options": "User", "width": 140},
        {"fieldname": "qty", "label": "Qty", "fieldtype": "Float", "width": 100},
        {"fieldname": "uom", "label": "UOM", "fieldtype": "Link", "options": "UOM", "width": 100},
        {"fieldname": "ordered_qty", "label": "Ordered Qty", "fieldtype": "Float", "width": 100},
        {"fieldname": "qty_to_order", "label": "Qty to Order", "fieldtype": "Float", "width": 100},
        {"fieldname": "purchase_order", "label": "Purchase Order", "fieldtype": "Link", "options": "Purchase Order", "width": 140, "align": "left"},
        # {"fieldname": "company", "label": "Company", "fieldtype": "Link", "options": "Company", "width": 140},
    ]

def get_data(filters):

    conditions = ""

    if filters.only_unassigned:
        conditions += " AND (mr_item.assigned_to IS NULL OR mr_item.assigned_to = '') "

    if filters.item_group:
        conditions += " AND mr_item.item_group = {0} ".format(frappe.db.escape(filters.item_group))

    if filters.request_type:
        conditions += " AND mr.material_request_type = {0} ".format(frappe.db.escape(filters.request_type))

    if filters.assigned_to:
        conditions += " AND mr_item.assigned_to = {0} ".format(frappe.db.escape(filters.assigned_to))

    query = """
SELECT 
    mr.name AS material_request,
    mr.material_request_type AS request_type,
    mr.transaction_date AS transaction_date,
    mr_item.schedule_date AS schedule_date,
    mr_item.name AS mr_item,
    mr_item.item_code AS item_code,
    SUM(IFNULL(mr_item.stock_qty, 0)) AS qty,
    IFNULL(mr_item.stock_uom, '') AS uom,
    SUM(IFNULL(mr_item.ordered_qty, 0)) AS ordered_qty, 
    (SUM(mr_item.stock_qty) - SUM(IFNULL(mr_item.ordered_qty, 0))) AS qty_to_order,
    mr_item.item_name AS item_name,
    mr_item.description AS description,
    item_default.default_supplier AS supplier,
    supplier.supplier_name AS supplier_name,
    mr_item.item_group AS item_group,
    mr_item.assigned_to AS assigned_to,
    mr.company AS company,
    po_item.parent AS purchase_order,
    po.status AS po_status
FROM
    `tabMaterial Request` mr
JOIN `tabMaterial Request Item` mr_item ON mr_item.parent = mr.name
JOIN `tabItem Default` item_default ON item_default.parent = mr_item.item_code AND item_default.company = mr.company
LEFT JOIN `tabSupplier` supplier ON supplier.name = item_default.default_supplier
LEFT JOIN `tabPurchase Order Item` po_item ON po_item.material_request_item = mr_item.name
LEFT JOIN `tabPurchase Order` po ON po.name = po_item.parent
WHERE
    mr.material_request_type = "Purchase"
    AND mr.docstatus = 1
    AND mr.status != "Stopped"
    {conditions}
GROUP BY mr.name, mr_item.item_code
HAVING
    SUM(IFNULL(mr_item.ordered_qty, 0)) < SUM(IFNULL(mr_item.stock_qty, 0))
ORDER BY mr.transaction_date ASC
    """.format(conditions=conditions)

    return frappe.db.sql(query, as_dict=1)